src/robot/variables/replacer.py

```diff
@@ -137,46 +137,59 @@ def _get_variable_item(self, match, value):
-                value = self._get_dict_variable_item(name, value, item)
-            elif is_list_like(value):
-                value = self._get_list_variable_item(name, value, item)
+            if is_sequence(value):
+                value = self._get_sequence_variable_item(name, value, item)
+            elif is_subscriptable(value):
+                value = self._get_subscriptable_variable_item(name, value, item)
             else:
                 raise VariableError(
-                    "Variable '%s' is %s, not list or dictionary, and thus "
-                    "accessing item '%s' from it is not possible."
-                    % (name, type_name(value), item)
+                    "Variable '%s' is %s, which is not subscriptable, and "
+                    "thus accessing item '%s' from it is not possible. To use "
+                    "'[%s]' as a literal value, it needs to be escaped like "
+                    "'\\[%s]'." % (name, type_name(value), item, item, item)
                 )
             name = '%s[%s]' % (name, item)
         return value
 
-    def _get_list_variable_item(self, name, variable, index):
+    def _get_sequence_variable_item(self, name, variable, index):
         index = self.replace_string(index)
         try:
-            index = self._parse_list_variable_index(index, name[0] == '$')
+            index = self._parse_sequence_variable_index(index, name[0] == '$')
         except ValueError:
-            raise VariableError("List '%s' used with invalid index '%s'."
-                                % (name, index))
+            raise VariableError("Sequence '%s' used with invalid index '%s'. "
+                                "To use '[%s]' as a literal value, it needs "
+                                "to be escaped like '\\[%s]'."
+                                % (name, index, index, index))
         try:
             return variable[index]
         except IndexError:
-            raise VariableError("List '%s' has no item in index %d."
+            raise VariableError("Sequence '%s' has no item in index %d."
                                 % (name, index))
 
-    def _parse_list_variable_index(self, index, support_slice=True):
+    def _parse_sequence_variable_index(self, index, support_slice=True):
         if ':' not in index:
             return int(index)
         if index.count(':') > 2 or not support_slice:
             raise ValueError
         return slice(*[int(i) if i else None for i in index.split(':')])
 
-    def _get_dict_variable_item(self, name, variable, key):
-        key = self.replace_scalar(key)
+    def _get_subscriptable_variable_item(self, name, variable, key):
+        if not isinstance(key, (int, slice)):
+            key = self.replace_scalar(key)
         try:
             return variable[key]
         except KeyError:
-            raise VariableError("Dictionary '%s' has no key '%s'."
+            raise VariableError("Subscriptable '%s' has no key '%s'."
```

Comment:
Yet another place where to use `type_name(variable, capitalize=True)`. In addition to that, perhaps it would be better to change `key` to `item`. Both ought to be OK with dictionaries, but the latter works better with custom objects.

Reply:
Done in d7110621c16a371e72b541b70f01cbea7f0f4c8e